mlflow/models/signature.py

```diff
@@ -0,0 +1,449 @@
+"""
+The :py:mod:`mlflow.models.signature` module provides an API for specification of model signature
+and of model example input.
+
+Model signature defines schema of model input and output. The schema is based on a simplified data
+frame model defined by a list of (optionally) named and (mandatory) typed columns. The currently
+supported data types are defined by :py:class:`DataType` and consists of basic scalar types and
+string and binary string. Neither arrays (and multidimensional arrays) nor structured types
+(struct, map) are supported at this time.
+
+Input example specifies one or more valid inputs (observations) to the model. The example is stored
+as json and so the examples are limited to jsonable content with the exception of binary data
+provided the signature is defined. If the example includes binary data and model signature is
+defined, the binary data is base64 encoded before producing output.
+
+User can construct schema by hand or infer it from a dataset. Input examples are provided as is
+(pandas.DataFrame, numpy.ndarray, dictinary or list) and are json serialized  and stored as part of
+the model artifacts.
+"""
+
+import base64
+import importlib
+import os
+from enum import Enum
+
+import json
+import numpy as np
+import pandas as pd
+from typing import List, Dict, TypeVar
+
+from mlflow.exceptions import MlflowException
+from mlflow.utils.proto_json_utils import NumpyEncoder
+
+
+class TensorsNotSupportedException(MlflowException):
+    def __init__(self, msg):
+        super().__init__("Multidimensional arrays (aka tensors) are not supported. "
+                         "{}".format(msg))
```

Comment:
just `msg`?

Reply:
Hmm, you mean dropping the shared part of the message altogether? Yeah I get it makes sense.